### PR TITLE
Add a dynamic inventory plugin for Online

### PR DESCRIPTION
**Pull request by remyleone**
##### SUMMARY

Add a dynamic inventory plugin for online

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

- online dynamic inventory

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (online_dynamic_inventory 04bf7e6a6b) last updated 2018/08/27 18:14:32 (GMT +200)
  config file = None
  configured module search path = [u'/Users/sieben/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/sieben/workspace/ansible/lib/ansible
  executable location = /Users/sieben/workspace/ansible/bin/ansible
  python version = 2.7.15 (default, Jun 17 2018, 12:46:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

### CODE REVIEW

File: lib/ansible/plugins/inventory/online.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+    def _fill_host_variables(self, hostname, host_infos):
+        targeted_attributes = (
+            "offer",
+            "id",
+            "hostname",
+            "location",
+            "boot_mode",
+            "power",
+            "boot_mode",
```
Comment (g1franc) [inline lib/ansible/plugins/inventory/online.py]: Two times boot_mode

**Comment by remyleone** [comment]
Nice to have features:

- Group by DC
- Group by RPN

File: lib/ansible/plugins/inventory/online.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
+# Copyright (c) 2017 Ansible Project
```
Comment (Spredzy) [inline lib/ansible/plugins/inventory/online.py]: s/2017/2018

File: lib/ansible/plugins/inventory/online.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+from ansible.module_utils.urls import open_url
+from ansible.plugins.inventory import BaseInventoryPlugin
+from ansible.module_utils._text import to_native, to_text
+from ansible.module_utils.six.moves.urllib.parse import urljoin
+
+
+def _fetch_information(url, token):
+    headers = {
+        'Authorization': "Bearer %s" % token,
+        'Content-type': 'application/json'
```
Comment (Spredzy) [inline lib/ansible/plugins/inventory/online.py]: suggestion: Adds an `User-Agent` header that identifies that its coming from Ansible. (I think we should do that also on the main modules). Not a blocker.

File: lib/ansible/plugins/inventory/online.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    }
+    try:
+        response = open_url(url, headers=headers)
+    except Exception as e:
+        raise AnsibleError("Error while fetching %s: %s" % (url, to_native(e)))
+
+    try:
+        raw_data = to_text(response.read(), errors='surrogate_or_strict')
```
Comment (Spredzy) [inline lib/ansible/plugins/inventory/online.py]: This seems to be related to https://github.com/ansible/ansible/pull/43715. Can we get a final agreement on the other PR first before pushing that here?

File: lib/ansible/plugins/inventory/online.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    def do_server_inventory(self, host_infos, hostname_preferences):
+
+        hostname = self._filter_host(host_infos=host_infos,
+                                     hostname_preferences=hostname_preferences)
+
+        # No suitable hostname were found in the attributes and the host won't be in the inventory
+        if not hostname:
+            return
+
+        self.inventory.add_host(host=hostname)
```
Comment (Spredzy) [inline lib/ansible/plugins/inventory/online.py:203]: I see there is no grouping in this inventory. Is this intentional for the first patch? Fine to me but I'd like to raise that grouping based on `os_name` of `offer` allow greater flexibility when dealing with a fleet of machine, might be interesting to add.

**Review by Spredzy — COMMENTED** [review]
Globally looks good to me. The only blocker is the line that reflects a change on https://github.com/ansible/ansible/pull/43715. I'd like us to have an agreement before pushing it.

File: lib/ansible/plugins/inventory/online.py (hunk 5 above)
Comment (remyleone) [reply lib/ansible/plugins/inventory/online.py:203]: I'm working on a grouping feature that could be easily extended to geographical location and private networking subnets.

Comment (remyleone) [reply lib/ansible/plugins/inventory/online.py:203]: Done for location and offer

**Comment by remyleone** [comment]
Done for RPN

**Comment by maxamillion** [comment]
shipit

**Comment by remyleone** [comment]
shipit

**Comment by maxamillion** [comment]
rebuild_merge

**Pull request merged**